### PR TITLE
refactor: migrate GasSettingDialog

### PR DESCRIPTION
**Pull request by lelenei**
## Description

Closes  mf-4344

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Review by UncleBill — DISMISSED** [review]

File: packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts [hunk 1]
```diff
@@ -22,18 +22,14 @@ export const useGasConfig = (gasLimit: number, minGasLimit: number) => {
...
-    useEffect(() => closeDialog, [closeDialog])
+    useEffect(() => GasSettingDialog.close, [])
 
     useEffect(() => {
         setGasLimit_(gasLimit)
     }, [gasLimit])
 
     useEffect(() => {
-        return WalletMessages.events.gasSettingDialogUpdated.on((evt) => {
+        return WalletMessages.events.gasSettingUpdated.on((evt) => {
```
Comment (UncleBill) [inline packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts]: Maybe you don't need this anymore.

File: packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts [hunk 2]
```diff
@@ -83,6 +79,6 @@ export const useGasConfig = (gasLimit: number, minGasLimit: number) => {
         gasConfig,
         gasLimit: gasLimit_,
         maxFee,
-        onCustomGasSetting: () => setGasSettingDialog({ open: true, gasLimit: gasLimit_, gasOption, minGasLimit }),
+        onCustomGasSetting: () => GasSettingDialog.open({ gasLimit: gasLimit_, gasOption, minGasLimit }),
```
Comment (UncleBill) [inline packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts]: Open and get setting result via `GasSettingDialog.openAndWaitForClose`

File: packages/shared/src/wallet/GasSettingBar.tsx [hunk 3]
```diff
@@ -49,7 +46,7 @@ export function GasSettingBar(props: GasSettingBarProps) {
 
     // set initial options
     useEffect(() => {
-        return WalletMessages.events.gasSettingDialogUpdated.on((evt) => {
+        return WalletMessages.events.gasSettingUpdated.on((evt) => {
```
Comment (guanbinrui) [inline packages/shared/src/wallet/GasSettingBar.tsx]: Added modal events emitter here: https://github.com/DimensionDev/Maskbook/pull/9868.

File: packages/dashboard/src/pages/Wallets/components/Transfer/TransferERC20.tsx [hunk 4]
```diff
@@ -138,7 +138,7 @@ export const TransferERC20 = memo<TransferERC20Props>(({ token }) => {
     )
     const { gasConfig, onCustomGasSetting, gasLimit, maxFee } = useGasConfig(gasLimit_, GAS_LIMIT)
 
-    const gasPrice = gasConfig.gasPrice || defaultGasPrice
+    const gasPrice = gasConfig?.gasPrice || defaultGasPrice
```
Comment (guanbinrui) [inline packages/dashboard/src/pages/Wallets/components/Transfer/TransferERC20.tsx]: No need ?

File: packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts [hunk 5]
```diff
@@ -69,7 +61,7 @@ export const useGasConfig = (gasLimit: number, minGasLimit: number) => {
         }
     }, [chainId, gasOptions?.normal])
 
-    const gasConfig = useMemo(() => {
+    const gasConfig: GasConfig = useMemo(() => {
```
Comment (guanbinrui) [inline packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts:64]: Use the generic parameter of useMemo.

File: packages/plugins/Wallet/src/messages.ts [hunk 6]
```diff
@@ -29,7 +29,7 @@ export type ConnectWalletDialogEvent =
           open: false
       }
 
-export type GasSettingDialogEvent = {
+export type GasSettingEvent = {
```
Comment (guanbinrui) [inline packages/plugins/Wallet/src/messages.ts]: Remove this type.

File: packages/plugins/Wallet/src/messages.ts [hunk 7]
```diff
@@ -64,7 +64,7 @@ export interface WalletMessage {
     /**
      * Gas setting dialog
      */
-    gasSettingDialogUpdated: GasSettingDialogEvent
+    gasSettingUpdated: GasSettingEvent
```
Comment (guanbinrui) [inline packages/plugins/Wallet/src/messages.ts]: Remove this event.

File: packages/shared/src/modals/GasSettingDialog/types.ts [hunk 8]
```diff
@@ -12,7 +12,7 @@ export interface GasSettingProps {
     gasLimit?: number
     minGasLimit?: number
     onGasLimitChange?: (newGasLimit: number) => void
-    gasOptionType?: GasOptionType
+    gasOptionType?: GasOptionType | null
```
Comment (guanbinrui) [inline packages/shared/src/modals/GasSettingDialog/types.ts]: Remove null.

File: packages/shared/src/modals/GasSettingDialog/index.tsx [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    const [gasOptionType, setGasOptionType] = useState<GasOptionType>(GasOptionType.NORMAL)
+    const [gasLimit, setGasLimit] = useState(0)
+    const [minGasLimit, setMinGasLimit] = useState(0)
+    const [gasPrice, setGasPrice] = useState<BigNumber.Value>(0)
+    const [maxFee, setMaxFee] = useState<BigNumber.Value>(0)
+    const [priorityFee, setPriorityFee] = useState<BigNumber.Value>(0)
+
+    const [open, dispatch] = useSingletonModal(ref, {
+        onOpen(props) {
+            setGasOptionType(props?.gasOption ?? GasOptionType.NORMAL)
```
Comment (guanbinrui) [inline packages/shared/src/modals/GasSettingDialog/index.tsx:32]: props is not undefined anymore.

File: packages/shared/src/modals/GasSettingDialog/index.tsx [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            setGasLimit(props?.gasLimit ?? 0)
+            setMinGasLimit(props?.minGasLimit ?? 0)
+            setGasPrice(props?.gasPrice ?? 0)
+            setMaxFee(props?.maxFee ?? 0)
+            setPriorityFee(props?.priorityFee ?? 0)
+        },
+        onClose(props) {
+            setGasOptionType(props?.gasOption ?? GasOptionType.NORMAL)
```
Comment (guanbinrui) [inline packages/shared/src/modals/GasSettingDialog/index.tsx:40]: The same here.

File: packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts (hunk 5 above)
Comment (lelenei) [reply packages/dashboard/src/pages/Wallets/hooks/useGasConfig.ts:64]: EIP1559GasConfig don't have a gastric

File: packages/dashboard/src/pages/Wallets/components/Transfer/TransferERC20.tsx (hunk 4 above)
Comment (lelenei) [reply packages/dashboard/src/pages/Wallets/components/Transfer/TransferERC20.tsx]: EIP1559GasConfig don't have a gastric

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**